docs/DesignDocs/libc_initialization.md

```diff
@@ -0,0 +1,172 @@
+This ensures that enclaves which link libc will call `_oe_init_c` during
+oe_call_init_functions.
+
+#### \_\_init\_tls
+
+OE initializes thread local storage via a call to `oe_thread_local_init` in
+enclave/core/linux/threadlocal.c which is called during `td_init()` which
+initializes a thread data structure.
+
+#### \_\_init\_ssp
+
+OE initializes the stack protector again during the `td_init()` function.
+
+#### Check for availability of stdin, stdout, and stderr
+
+OE enclaves do console I/O by making ocalls to the host. Because of this,
+the job of checking console I/O file descriptors is handled by the host's
+libc implementation.
+
+#### Initialize global variables
+
+* \_\_environ
+
+    \_\_environ is a pointer to the `envp` argument that can be passed to main, which
+    contains the current environment variables. Because the enclave does not inherit an
+    initial set of environment variables, this is not initialized. Environment-based libc
+    functions such as `putenv`, `getenv`, etc. will still work as expected. No musl functions
+    assume __environ is non-null.
+
+* \_\_hwcap
+
+    \_\_hwcap is a bitmask of flags detailing whether various features are supported
+    by the system's CPU. In general, hwcaps are used by the kernel, but musl uses
+    a few of them for threading. It checks for `CPU_HAS_CAS_L`, `CPU_HAS_LLSC`, and
+    `HWCAP_TLS`.
+
+    Each of these are used for thread local storage. Because OE implements its own
+    TLS, these flags do not need to be initialized.
+
+* \_\_sysinfo
+
+    On x86_64 this is not used by musl. OE does not need to initialize.
+
+* \_\_progname
+
+    \_\_progname contains the name of the executing program (`argv[0]`). It is used by
+    musl libc in warn and err functions to print a message to stderr prefixed by the
+    program name. Musl initializes __progname to null, so these calls will succeed and
+    simply have an empty prefix.
+
+    OE does not need to initialize this value explicitly.
+
+#### Initialize members of the libc struct
+
+Musl libc maintains a libc struct which tracks some global configurations. `__init_libc`
+initializes some of these members.
+
+* auxv
+
+    This is a pointer to the part of the auxilary vector that is not used for argc, argv,
+    and envp. In general, the auxillary vector is a mechanism for the kernel's loader
+    to pass information to a userspace program when it is executed. Musl libc uses the
+    auxillary vector for the following:
+
+    * Thread local storage
+    * Dynamic loading of shared objects.
+    * The `getauxval` call used for reading from the auxilary vector by userspace program.
+
+    OE initializes auxv as an array containing 1 zero entry. This denotes an empty vector.
+
+* page\_size
+
+    Musl libc uses page_size for dynamic linking, pthreads, memory allocation, and the
+    sysconf call.
+
+    In order to support calls to sysconf, OE initializes page\_size
```

Comment:
Nit: Missing period at end of sentence. Probably also useful to indicate what we initialize it to (as done for `secure` below).